examples/htool_i2c.c

```diff
@@ -0,0 +1,220 @@
+          printf("0x%02X ", (i * 8 + b));
+        }
+      }
+    }
+    printf("\n");
+  }
+
+  return 0;
+}
+
+static int i2c_read(struct libhoth_device *dev,
+                    const struct htool_invocation *inv) {
+  uint32_t bus;
+  uint32_t freq;
+  uint32_t addr;
+  uint32_t offset;
+  uint32_t length;
+
+  if (htool_get_param_u32(inv, "bus", &bus) ||
+      htool_get_param_u32(inv, "frequency", &freq) ||
+      htool_get_param_u32(inv, "address", &addr) ||
+      htool_get_param_u32(inv, "offset", &offset) ||
+      htool_get_param_u32(inv, "length", &length)) {
+    return -1;
+  }
+
+  struct ec_request_i2c_transfer request;
+  request.bus_number = (uint8_t)(bus & 0xFF);
+  request.dev_address = (uint8_t)(addr & 0x7F);
+  request.size_read = (uint16_t)(length & 0xFFFF);
+  request.speed_khz = (uint16_t)(freq & 0xFFFF);
+  request.flags = 0;
+
+  if (offset == UINT32_MAX) {
+    request.size_write = 0;
+  } else {
+    request.size_write = 1;
+    request.arg_bytes[1] = (offset & 0xFF);
+  }
+
+  uint8_t response[sizeof(struct ec_response_i2c_transfer)];
+  size_t rLen = 0;
+  int ret = htool_exec_hostcmd(
+      dev, EC_CMD_BOARD_SPECIFIC_BASE + EC_PRV_CMD_HOTH_I2C_TRANSFER, 0,
+      &request, sizeof(request), &response, sizeof(response), &rLen);
+  if (ret != 0) {
+    fprintf(stderr, "HOTH_I2C_TRANSFER [read] error code: %d\n", ret);
+    return -1;
+  }
+  if (rLen != sizeof(response)) {
+    fprintf(stderr,
+            "HOTH_I2C_TRANSFER [read] expected exactly %ld response "
+            "bytes, got %ld\n",
+            sizeof(response), rLen);
+    return -1;
+  }
+
+  struct ec_response_i2c_transfer *pI2t =
+      (struct ec_response_i2c_transfer *)(response);
+
+  printf("Read %u bytes from 0x%02x\n", pI2t->read_bytes, request.dev_address);
+  for (uint16_t i = 0; i < pI2t->read_bytes; i++) {
+    printf("0x%02X ", pI2t->resp_bytes[i]);
+  }
+  printf("\n");
+
+  return 0;
+}
+
+static int i2c_write(struct libhoth_device *dev,
+                     const struct htool_invocation *inv) {
+  uint32_t bus;
+  uint32_t freq;
+  uint32_t addr;
+  bool no_stop;
+  char *byte_stream = NULL;
+
+  if (htool_get_param_u32(inv, "bus", &bus) ||
+      htool_get_param_u32(inv, "frequency", &freq) ||
+      htool_get_param_u32(inv, "address", &addr) ||
+      htool_get_param_bool(inv, "no_stop", &no_stop) ||
+      htool_get_param_string(inv, "byte_stream", (const char **)&byte_stream)) {
+    return -1;
+  }
+
+  if (byte_stream == NULL) {
+    return -1;
+  }
+
+  struct ec_request_i2c_transfer request;
+  request.bus_number = (uint8_t)(bus & 0xFF);
+  request.dev_address = (uint8_t)(addr & 0x7F);
+  request.speed_khz = (uint16_t)(freq & 0xFFFF);
+  request.flags = (no_stop ? I2C_BITS_NO_STOP : 0) | I2C_BITS_WRITE;
+  request.size_read = 0;
+
+  uint16_t idx = 0;
+  char *tk = strtok(byte_stream, " ");
+  while (tk && (idx < I2C_TRANSFER_DATA_MAX_SIZE_BYTES)) {
+    unsigned long int parsed = strtoul(tk, NULL, 0);
```

Comment:
`strtoul` will silently return `0` if it can't perform the conversion. We should check for and report this failure.

Reply:
Right, we probably want to do this or a similar check. If these commands are going to be used manually, then someone may accidentally type an invalid byte that `strtoul` silently converts to a zero.

Reply:
Checking the output pointer now as well as errno